server/src-lib/Hasura/Server/Query.hs

```diff
@@ -369,7 +381,15 @@ runQueryM rq =
 
       RQRunSql q                   -> runRunSQL q
 
-      RQBulk qs                    -> encJFromList <$> indexedMapM runQueryM qs
+      RQBulk qs                    -> unless (allSameTxAccess qs) throwTxMismatch >>
+                                      encJFromList <$> indexedMapM  runQueryM qs
+      where
+        allSameTxAccess qs = allReadWrite qs || allReadOnly qs
+        allReadWrite = all isReadWrite
+        allReadOnly = all isReadOnly
+        isReadWrite q = getQueryAccessMode q == Just Q.ReadWrite || getQueryAccessMode q == Nothing
+        isReadOnly q = getQueryAccessMode q == Just Q.ReadOnly
+        throwTxMismatch = throw400 BadRequest "bulk query has transaction access mode mismatch"
```

Comment:
How hard would it be to provide slightly more information in this error message? I’m envisioning something like:

```
incompatible access mode requirements in bulk query:
  run_sql at $.query[0] requires read-only access
  track_table at $.query[1] requires read-write access
```

It wouldn’t have to list all the contradictions, just the first one.